cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.NamespaceNotFoundException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.config.PreferencesService;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.proto.PreferencesDetail;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * Program Preferences HTTP Handler for internal usage
+ */
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3)
+public class PreferencesHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+
+  private static final Gson GSON = new Gson();
+
+  private final PreferencesService preferencesService;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+
+  @Inject
+  PreferencesHttpHandlerInternal(PreferencesService preferencesService,
+                                 ApplicationLifecycleService applicationLifecycleService,
+                                 NamespaceQueryAdmin namespaceQueryAdmin) {
+    this.preferencesService = preferencesService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+  }
+
+  @Path("/preferences")
+  @GET
+  public void getInstancePreferences(HttpRequest request, HttpResponder responder) {
+    PreferencesDetail detail = preferencesService.getPreferences();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail, PreferencesDetail.class));
```

Comment:
I believe sendJson will serialize the object to json automatically, which mean you don't need the GSON.toJson()

Reply:
void sendJson(HttpResponseStatus status, String jsonString);

Don't we still need GSON.toJson() to convert to jsonString?

Reply:
oh this is the netty responder, got this mixed up with HttpServiceResponder. Ignore my comment.

Reply:
Resolved.